### PR TITLE
[WIP] For test case

### PR DESCRIPTION
**Pull request by Jay357089**
issue #126 issue #81 
to support pre-defined dict from external column file, we add Key word COLUMNDICT in option of load DML to define the column name and column file path as below:

LOAD DATA inpath 'D:/inputPath/carbonTest.csv' INTO TABLE carbonTest options( 'COLUMNDICT '='name:D:/Carbon/carbonData/name.csv,sex:D:/Carbon/carbonData/sex.csv')

user needs to set as "column_Name: column_Path", and multiple columns and file paths can be delimited by ",'.

we handle the above DML as below:
1. in LoadCube, set column name and dictionanry file path in carbonLoadModel.
2. get the column name and column file path from carbonLoadModel, test whether the column is exists, then generate new dict using the column file.
3. read the old dictionary file and compare with new dict and remove duplication, then write dict file.
4 if the column data is also provided by fact table, then drop the column data from fact table when generating dict using fact table file.

### CODE REVIEW

**Comment by Jay357089** [comment]
@QiangCai @ravipesala pls review

File: integration/spark/src/main/scala/org/apache/spark/sql/CarbonSqlParser.scala [hunk 1]
```diff
@@ -1041,6 +1042,7 @@ class CarbonSqlParser()
   protected lazy val partitionOptions: Parser[(String, String)] =
     ((DELIMITER ~ stringLit) | (QUOTECHAR ~ stringLit) | (FILEHEADER ~ stringLit) |
       (ESCAPECHAR ~ stringLit) | (MULTILINE ~ stringLit) |
+      (COLUMNDICT ~ stringLit) |
```
Comment (QiangCai) [inline integration/spark/src/main/scala/org/apache/spark/sql/CarbonSqlParser.scala]: better to not wrap line

File: integration/spark/src/main/scala/org/carbondata/spark/util/GlobalDictionaryUtil.scala [hunk 2]
```diff
@@ -433,16 +442,92 @@ object GlobalDictionaryUtil extends Logging {
...
+      if (!dimensions.exists(_.getColName.equalsIgnoreCase(colName))) {
+        logError(s"No column $colName exists in $dataBaseName.$tableName")
+        throw new IllegalArgumentException(s"No column $colName exists " +
+          s"in $dataBaseName.$tableName")
+      }
+      breakable {
+        dimensions.foreach(x =>
+          if (x.getColName.equalsIgnoreCase(colName)) {
+            dictCarbonDimensions += x
+            dictColumnPaths += colPathMapTrim.substring(colName.length + 1)
```
Comment (QiangCai) [inline integration/spark/src/main/scala/org/carbondata/spark/util/GlobalDictionaryUtil.scala]: use FileUtils.getPaths to convert path

File: integration/spark/src/main/scala/org/carbondata/spark/util/GlobalDictionaryUtil.scala [hunk 3]
```diff
@@ -433,16 +442,92 @@ object GlobalDictionaryUtil extends Logging {
...
+   */
+  private def generateExternalColumnDictionary(colDictFilePath: String,
+                                               table: CarbonTableIdentifier,
+                                               dimensions: Array[CarbonDimension],
+                                               carbonLoadModel: CarbonLoadModel,
+                                               sqlContext: SQLContext,
+                                               hdfsLocation: String,
+                                               dictFolderPath: String) = {
+    // get external dictionary column
+    val extColumns = getExternalColumnDictPath(colDictFilePath, table, dimensions)
```
Comment (QiangCai) [inline integration/spark/src/main/scala/org/carbondata/spark/util/GlobalDictionaryUtil.scala]: (xxx1, xxx2)

File: integration/spark/src/test/scala/org/carbondata/spark/util/ExternalColumnDictionaryTestCase.scala [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    carbonLoadModel
+  }
+
+  override def beforeAll {
+    buildTestData
+    buildTable
+    buildRelation
+  }
+
+  def checkDictionary(relation: CarbonRelation, columnName: String, value: String){
```
Comment (QiangCai) [inline integration/spark/src/test/scala/org/carbondata/spark/util/ExternalColumnDictionaryTestCase.scala]: please extract this method to a util object. better to reuse code.

File: integration/spark/src/main/scala/org/carbondata/spark/util/GlobalDictionaryUtil.scala [hunk 5]
```diff
@@ -61,21 +61,23 @@ object GlobalDictionaryUtil extends Logging {
...
     val dimensionBuffer = new ArrayBuffer[CarbonDimension]
     val columnNameBuffer = new ArrayBuffer[String]
     val dimensionsWithDict = dimensions.filter(hasEncoding(_, Encoding.DICTIONARY,
       Encoding.DIRECT_DICTIONARY))
     dimensionsWithDict.foreach { dim =>
       breakable {
         headers.zipWithIndex.foreach { h =>
-          if (dim.getColName.equalsIgnoreCase(h._1)) {
+          if (dim.getColName.equalsIgnoreCase(h._1) && (extDimensions.isEmpty ||
+            !extDimensions.get.exists(_.getColName.equalsIgnoreCase(dim.getColName)))) {
```
Comment (QiangCai) [inline integration/spark/src/main/scala/org/carbondata/spark/util/GlobalDictionaryUtil.scala]: for complex type column, need check whether all primitive column involved by extDimensions.

Comment (Jay357089) [inline integration/spark/src/main/scala/org/carbondata/spark/util/GlobalDictionaryUtil.scala]: @gvramana  for David's question, whether we need to support that user specify some primitive columns in complex type to make global dictionary by using predefined dictionary, while other primitive columns not ? or user can only specify the whole complex type column? please response.

Comment (gvramana) [inline integration/spark/src/main/scala/org/carbondata/spark/util/GlobalDictionaryUtil.scala]: User cannot specify, whole complex type column, he has to specify primitive
columns inside complex type as predefined.

**Comment by Jay357089** [comment]
@QiangCai  pls review seriously.

File: integration/spark/src/main/scala/org/carbondata/spark/util/GlobalDictionaryUtil.scala [hunk 6]
```diff
@@ -91,7 +93,11 @@ object GlobalDictionaryUtil extends Logging {
   }
 
   /**
-   * use this method to judge whether CarbonDimension use some encoding or not
+   *  use this method to judge whether CarbonDimension use some encoding or not
```
Comment (QiangCai) [inline integration/spark/src/main/scala/org/carbondata/spark/util/GlobalDictionaryUtil.scala]: just need one blank

File: integration/spark/src/main/scala/org/carbondata/spark/util/GlobalDictionaryUtil.scala [hunk 7]
```diff
@@ -65,13 +67,13 @@ object GlobalDictionaryUtil extends Logging {
...
-   * @param columns    column list of csv file
-   * @return java.lang.String[]
+   * @param dimensions  dimension list of schema
+   * @param headers  column headers
+   * @param columns column list of csv file
    */
   def pruneDimensions(dimensions: Array[CarbonDimension],
-      headers: Array[String],
-      columns: Array[String]): (Array[CarbonDimension], Array[String]) = {
+                      headers: Array[String],
```
Comment (QiangCai) [inline integration/spark/src/main/scala/org/carbondata/spark/util/GlobalDictionaryUtil.scala]: check code style

File: integration/spark/src/main/scala/org/carbondata/spark/util/GlobalDictionaryUtil.scala [hunk 8]
```diff
@@ -109,24 +115,30 @@ object GlobalDictionaryUtil extends Logging {
...
-        gatherDimensionByEncoding(c, encoding, excludeEncoding, dimensionsWithEncoding)
+        gatherDimensionByEncoding(c, encoding, excludeEncoding,
+          dimensionsWithEncoding, forPreDefDict)
       }
     } else {
       if (dimension.hasEncoding(encoding) &&
           (excludeEncoding == null || !dimension.hasEncoding(excludeEncoding))) {
-        dimensionsWithEncoding += dimension
+        if ((forPreDefDict && dimension.getHasPredefDict) ||
+          (!forPreDefDict && !dimension.getHasPredefDict)) {
```
Comment (QiangCai) [inline integration/spark/src/main/scala/org/carbondata/spark/util/GlobalDictionaryUtil.scala]: only need !forPreDefDict

File: integration/spark/src/main/scala/org/carbondata/spark/util/GlobalDictionaryUtil.scala [hunk 9]
```diff
@@ -109,24 +115,30 @@ object GlobalDictionaryUtil extends Logging {
...
+          dimensionsWithEncoding += dimension
+        }
       }
     }
   }
 
-  def getPrimDimensionWithDict(dimension: CarbonDimension): Array[CarbonDimension] = {
+  def getPrimDimensionWithDict(dimension: CarbonDimension,
+                               forPreDefDict: Boolean): Array[CarbonDimension] = {
```
Comment (QiangCai) [inline integration/spark/src/main/scala/org/carbondata/spark/util/GlobalDictionaryUtil.scala]: code sytle

File: core/src/main/java/org/carbondata/core/carbon/metadata/schema/table/column/CarbonDimension.java [hunk 10]
```diff
@@ -134,4 +134,31 @@ public int getColumnGroupOrdinal() {
     }
     return true;
   }
+
+  /**
+   *  indicates whether the dimension has predefined dictionary
+   */
+  private boolean hasPredefDict = false;
```
Comment (QiangCai) [inline core/src/main/java/org/carbondata/core/carbon/metadata/schema/table/column/CarbonDimension.java]: These variables are only using for data loading.
It it not good to add these variables in CarbonDimension.

File: integration/spark/src/main/scala/org/carbondata/spark/util/GlobalDictionaryUtil.scala [hunk 11]
```diff
@@ -445,16 +461,125 @@ object GlobalDictionaryUtil extends Logging {
...
+          if (parentDimName.isEmpty) colName.split("\\.")(0)
+          else "." + colName.split("\\.")(0)
+      }
+    }
+    // judge whether the column is exists
+    val preDictDimensionOption = dimensions.filter(
+      _.getColName.equalsIgnoreCase(dimParent))
+    if (preDictDimensionOption.length == 0) {
+      logError(s"No column $dimParent exists in ${table.getDatabaseName}.${table.getTableName}")
+      throw new IllegalArgumentException(s"No column $colName exists " +
```
Comment (QiangCai) [inline integration/spark/src/main/scala/org/carbondata/spark/util/GlobalDictionaryUtil.scala]: use DataLoadingException

File: integration/spark/src/main/scala/org/carbondata/spark/util/GlobalDictionaryUtil.scala [hunk 12]
```diff
@@ -445,16 +461,125 @@ object GlobalDictionaryUtil extends Logging {
...
+  def setPredefineDict(dimensions: Array[CarbonDimension],
+                       table: CarbonTableIdentifier,
+                       colName: String,
+                       colDictPath: String,
+                       parentDimName: String = "") {
+    val dimParent = parentDimName + {
+      colName match {
+        case "" => colName
+        case _ =>
+          if (parentDimName.isEmpty) colName.split("\\.")(0)
```
Comment (QiangCai) [inline integration/spark/src/main/scala/org/carbondata/spark/util/GlobalDictionaryUtil.scala]: { if (parentDimName.isEmpty) "" esle "."  } +  colName.split("\\.")(0)

Comment (QiangCai) [inline integration/spark/src/main/scala/org/carbondata/spark/util/GlobalDictionaryUtil.scala]: set colName.split("\\.") to a local variable

File: integration/spark/src/main/scala/org/carbondata/spark/rdd/CarbonGlobalDictionaryRDD.scala [hunk 13]
```diff
@@ -321,3 +326,83 @@ class CarbonGlobalDictionaryGenerateRDD(
...
+    result
+  }
+
+  override def compute(split: Partition, context: TaskContext)
+  : Iterator[(Int, ColumnDistinctValues)] = {
+    val theSplit = split.asInstanceOf[CarbonColumnDictPatition]
+    val primDimension = theSplit.preDefDictDimension
+    // read the column dict data
+    val preDefDictFilePath = primDimension.getPreDefDictFilePath
+    val inputStream = FileFactory.getDataInputStream(preDefDictFilePath,
```
Comment (QiangCai) [inline integration/spark/src/main/scala/org/carbondata/spark/rdd/CarbonGlobalDictionaryRDD.scala]: need to release inputStream in finally

**Comment by Jay357089** [comment]
@QiangCai handled all the comments. pls review again.

File: integration/spark/src/main/scala/org/carbondata/spark/rdd/CarbonGlobalDictionaryRDD.scala [hunk 14]
```diff
@@ -321,3 +327,91 @@ class CarbonGlobalDictionaryGenerateRDD(
...
+    } catch {
+      case _ =>
+        throw new DataLoadingException("Error in reading pre-defined dictionary file path.")
+    }
+    val csvReader = new CSVReader(new InputStreamReader(inputStream, Charset.defaultCharset),
+      CSVReader.DEFAULT_SKIP_LINES, new CSVParser())
+    // read the column data to list
+    val colDict = csvReader.readAll.iterator
+    try {
+      csvReader.close()
```
Comment (QiangCai) [inline integration/spark/src/main/scala/org/carbondata/spark/rdd/CarbonGlobalDictionaryRDD.scala]: suggest to use finally

**Comment by Jay357089** [comment]
retest this please

**Comment by Jay357089** [comment]
This pr is just used to run CI test. any comments, pls review on https://github.com/apache/incubator-carbondata/pull/16

**Comment by ravipesala** [comment]
@Jay357089 Are you sure this PR is not required for version? Can you please confirm with @gvramana and @sraghunandan

**Comment by Jay357089** [comment]
@ravipesala  do not get your point. This one is a AR, and should be handled.  you can check issue#126 and issue#81 which assigned by Raghu and Ramana. i have raise a new PR in https://github.com/apache/incubator-carbondata/pull/16  , and the PR here is just used to run CI test. thanks.

**Comment by Jay357089** [comment]
@ravipesala you mean that PR raised here will be merged to version, while raised in apache/incubator-carbondata will not be merged to version? if it is like this, then i should raise PR here.

**Pull request closed without merging**